Assignment.Tests/PingProcessTests.cs

```diff
@@ -0,0 +1,260 @@
+    string LimitPingArg { get; set; }
+    string PingOutputLikeExpression {get; set;}
+#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+    [TestInitialize]
+    public void TestInitialize()
+    {
+        IsUnix = Environment.OSVersion.Platform is PlatformID.Unix;
+        (string arg, string exp)  = IsUnix ? ("-c", PingOutputLikeExpressionUnix) : ("-n", PingOutputLikeExpressionWindows);
+        LimitPingArg = arg;
+        PingOutputLikeExpression = exp;
+        Sut = new();
+    }
+
+    [TestMethod]
+    public void Start_PingProcess_Success()
+    {
+        Process process = Process.Start("ping", $"{LimitPingArg} 4 localhost");
+        process.WaitForExit();
+        Assert.AreEqual<int>(0, process.ExitCode);
+    }
+
+    [TestMethod]
+    public void Run_GoogleDotCom_Success()
+    {
+
+        int expectedExitCode = Environment.GetEnvironmentVariable("GITHUB_ACTIONS") is not null || IsUnix ? 1 : 0;
+
+        int exitCode = Sut.Run("google.com").ExitCode;
+        Assert.AreEqual<int>(expectedExitCode, exitCode);
+    }
+    
+
+    [TestMethod]
+    public void Run_LocalHost_Success()
+    {
+        int exitCode = Sut.Run("localhost").ExitCode;
+        Assert.AreEqual<int>(0, exitCode);
+    }
+    //When running this test, stdOutput comes back as Empty
+    [TestMethod]
+    public void Run_InvalidAddressOutput_Success()
+    {
+
+        (string expectedOutput, int expectedExitCode)  = IsUnix ? ("ping: badaddress: Temporary failure in name resolution", 2) : ("Ping request could not find host badaddress. Please check the name and try again.",1);
+        (int exitCode, string? stdOutput, string? stdError) = Sut.Run("badaddress");
+        string actualOutput = IsUnix ? stdError! : stdOutput!;
+        //In Unix, error is logged to StdError, not StdOutput
+        Assert.IsFalse(string.IsNullOrWhiteSpace(actualOutput));
+        actualOutput = WildcardPattern.NormalizeLineEndings(actualOutput!.Trim());
+        Assert.AreEqual<string?>(expectedOutput, actualOutput, $"Output is unexpected: {stdOutput}");
+        // 2 is the exit code for invalid address in Unix
+        // 1 is the exit code for invalid address in Windows
+        Assert.AreEqual<int>(expectedExitCode, exitCode);
+    }
+
+
+    [TestMethod]
+    public void Run_CaptureStdOutput_Success()
+    {
+        PingResult result = Sut.Run("localhost");
+        AssertValidPingOutput(result);
+    }
+
+    [TestMethod]
+    public void RunTaskAsync_Success()
+    {
+        // Do NOT use async/await in this test.
+        Task<PingResult> result = Sut.RunTaskAsync("localhost");
+        AssertValidPingOutput(result.Result);
+    }
+
+    [TestMethod]
+    async public Task RunTaskAsync_WithProgress_Success()
+    {
+        string progressAppended = "";
+        var progress = new Progress<string?>((output) => progressAppended += output + Environment.NewLine);
```

Comment:
```suggestion
        var progress = new PingProccess<string?>((output) => progressAppended += output + Environment.NewLine);
```
`Progress` should be `PingProgress`.  When I did the last revert, I ended up removing this line. The test passed when it is not consistently passing when we worked on it. The reason why is because the order of progress will may not be ordered when using the regular `Report` method from `Progress` class. This was because of how the `SynchronizationContext` worked with the `Report` method. My understanding is that Report is async so it whichever thread returns its report is not guaranteed to be in a set order due to the Sync Context being related to the thread.

This was remedied by instead deriving `IProgress` in `PingProgress`. Using PingProgress will fix issue with inconsistent testing results.